resolwe/flow/models/utils/duplicate.py

```diff
@@ -0,0 +1,301 @@
+                entity.collection = collection_mapping[entity.collection.pk]
+        else:
+            entity.collection = None
+
+        return entity
+
+    return processor
+
+
+def process_data(
+    inherit_collection, inherit_entity, collection_mapping, entity_mapping
+):
+    def processor(datum):
+        if inherit_collection:
+            if datum.collection.id in collection_mapping:
+                datum.collection = collection_mapping[datum.collection.pk]
+        else:
+            datum.collection = None
+
+        if inherit_entity:
+            if datum.entity.id in entity_mapping:
+                datum.entity = entity_mapping[datum.entity.pk]
+        else:
+            datum.entity = None
+
+        return datum
+
+    return processor
+
+
+def copy_objects(objects, contributor, name_prefix, obj_processor=None):
+    """"""
+    first = objects.first()
+    if not first:
+        return objects
+
+    name_max_length = first._meta.get_field('name').max_length
+    content_type = ContentType.objects.get_for_model(first)
+    model = first._meta.model
+
+    new_objects = []
+    for obj in objects:
+        new_obj = deepcopy(obj)
+        new_obj.pk = None
+        new_obj.slug = None
+        new_obj.contributor = contributor
+        new_obj.name = "{} {}".format(name_prefix, obj.name)
+
+        if len(new_obj.name) > name_max_length:
+            new_obj.name = "{}...".format(new_obj.name[: name_max_length - 3])
+
+        if obj_processor:
+            new_obj = obj_processor(new_obj)
+
+        new_objects.append(new_obj)
+
+    try:
+        # Add another atomic block to avoid corupting the main one.
+        with transaction.atomic():
+            model.objects.bulk_create(new_objects)
+    except IntegrityError:
+        # Probably a slug collision occured, try to create objects one by one.
+        for obj in new_objects:
+            obj.slug = None
+            obj.save()
+
+    for old, new in zip(objects, new_objects):
+        new.created = old.created
+        new.duplicated = timezone.now()
+
+    model.objects.bulk_update(new_objects, ['created', 'duplicated'])
+
+    pk_mapping = {old.pk: new.pk for old, new in zip(objects, new_objects)}
+    _bulk_copy_permissions(content_type, pk_mapping)
+    _bulk_assign_contributor_permission(content_type, pk_mapping.values(), contributor)
+
+    return new_objects
+
+
+@transaction.atomic
+def bulk_duplicate(
+    collections=None,
+    entities=None,
+    data=None,
+    contributor=None,
+    inherit_collection=False,
+    inherit_entity=False,
+    name_prefix=None,
+):
+    """Rewire data inputs, dependencies and save models."""
+    # Prevent circular import.
+    from resolwe.flow.models import (
+        Collection,
+        Data,
+        DataDependency,
+        DataMigrationHistory,
+        Entity,
+    )
+
+    name_prefix = name_prefix or "Copy of"
+
+    if contributor is None:
+        raise ValueError("Contributor must be give.")
```

Comment:
`must be given. `

Maybe you could just use `if contributor` syntax to make it consistent with the rest of the code.